### PR TITLE
Use stemming and tfidf to find the best match for what user utters.  …

### PR DESCRIPTION
**Pull request by tamarahills**
…Use Amazon.SearchQuery to get more of the utterance.

### CODE REVIEW

File: state_handlers.js [hunk 1]
```diff
@@ -75,51 +91,198 @@ var state_handlers = {
...
+        delete this.attributes['State'];
+        this.emit(':saveState', true);
+      },
+      Unhandled: function() {
+        console.log(
+          'TITLES_DECISION_MODE:Unhandled: ' + this.event.request.intent.name
+        );
+
+        var message =
+          'Sorry, unhandled intent: in state: ' +
```
Comment (dylano) [inline state_handlers.js]: String

File: state_handlers.js [hunk 2]
```diff
@@ -193,6 +382,40 @@ function getTitlesHelper(stateObj) {
...
+  ) {
+    console.log('article title: ' + articleJson[i].title);
+    let num = i + 1;
+    retSpeech +=
+      num +
+      '.  ' +
+      articleJson[i].title +
+      '. ' +
+      articleJson[i].lengthMinutes +
+      ' minutes.  ';
```
Comment (dylano) [inline state_handlers.js]: Just a style thing but I think templates make this kind of string construction easier to read:
```
retSpeech = `${retSpeech} ${i+1}. ${articleJson[i].title}. ${articleJson[i].lengthMinutes} minutes.  `
```

File: state_handlers.js [hunk 3]
```diff
@@ -193,6 +382,40 @@ function getTitlesHelper(stateObj) {
...
+  let articleLen = Object.keys(articleJson).length;
+  console.log('Length of array is: ' + articleLen);
+  let arrArticles = Object.keys(articleJson);
+  let retSpeech = '<break />';
+  let j = 0;
+  console.log(
+    'stateObj.attributes[titleCount] is: ' + stateObj.attributes['titleCount']
+  );
+  for (
+    let i = stateObj.attributes['titleCount'];
```
Comment (dylano) [inline state_handlers.js]: I haven't taken a swing at an alternate implementation but this loop feels like it's more complex than it needs to be for the purpose of 'grab the next X titles'. Maybe a `slice()` and a `for(... of ...)` would simplify it and allow you to eliminate the multiple counters.

File: state_handlers.js [hunk 4]
```diff
@@ -25,36 +28,49 @@ var state_handlers = {
...
+
         var message = 'Welcome to Scout. You can say, get titles to begin.';
         this.response.speak(message).listen(message);
         this.emit(':responseReady');
       },
       'AMAZON.StopIntent': function() {
-        var message = 'Good bye.';
-        this.response.speak(message);
+        console.log('START_MODE:AMAZON.StopIntent');
+        this.response.speak(ALEXA_STOP_RESP);
```
Comment (dylano) [inline state_handlers.js]: Need to scope this one `constants.errString.ALEXA_STOP_RESP`

File: constants.js [hunk 5]
```diff
@@ -17,8 +17,41 @@ module.exports = Object.freeze({
   states: {
     START_MODE: '',
     PLAY_MODE: '_PLAY_MODE',
-    RESUME_DECISION_MODE: '_RESUME_DECISION_MODE'
+    RESUME_DECISION_MODE: '_RESUME_DECISION_MODE',
+    TITLES_DECISION_MODE: '_TITLES_DECISION_MODE'
   },
+  errString: {
```
[the remark this answers is not in the document]
Comment (tamarahills) [reply constants.js]: Yes, strings sounds good for now.

File: constants.js [hunk 6]
```diff
@@ -17,8 +17,41 @@ module.exports = Object.freeze({
...
+    TITLES_REPEAT: 'Repeating last titles<break />',
+    TITLE_CHOOSE_SUMM_FULL:
+      'Would you like to hear a summary or the full article of ',
+    TITLE_CHOICE_REPROMPT:
+      'You can say summary to hear a summary or full article to hear the full article',
+    ARTICLE_FAIL_MSG:
+      'Unable to find the article.  Please try' +
+      ' saying get titles to hear your titles.',
+    ALEXA_STOP_RESP: 'Goodbye!  Thanks for listening to the web with Scout'
+  },
```
[the remark this answers is not in the document]
Comment (tamarahills) [reply constants.js]: Yes I'll file a follow up.  We are going to need this for eventual translations too.

File: state_handlers.js (hunk 4 above)
Comment (tamarahills) [reply state_handlers.js]: good catch.

File: state_handlers.js [hunk 7]
```diff
@@ -25,36 +28,49 @@ var state_handlers = {
...
         this.emit(':responseReady');
       },
       'AMAZON.CancelIntent': function() {
+        console.log('START_MODE:AMAZON.CancelIntent');
         var message = 'Good bye.';
```
[the remark this answers is not in the document]
Comment (tamarahills) [reply state_handlers.js]: fixed.

File: state_handlers.js [hunk 8]
```diff
@@ -25,36 +28,49 @@ var state_handlers = {
...
       },
       SessionEndedRequest: function() {
+        console.log('START_MODE:AMAZON.SessionEndedRequest');
         // No session ended logic
       },
       Unhandled: function() {
+        console.log('START_MODE:Unhandled: ' + this.event.request.intent.name);
         var message =
           'Sorry, I could not understand. Please say, get titles to hear your titles.';
         this.response.speak(message).listen(message);
```
[the remark this answers is not in the document]
Comment (tamarahills) [reply state_handlers.js]: fixed

File: state_handlers.js [hunk 9]
```diff
@@ -65,7 +81,7 @@ var state_handlers = {
   playModeIntentHandlers: Alexa.CreateStateHandler(constants.states.PLAY_MODE, {
     //Intent handlers for PLAY_MODE
     LaunchRequest: function() {
-      console.log('PLAY_MODE|LaunchRequest');
+      console.log('PLAY_MODE:LaunchRequest');
       this.handler.state = constants.states.START_MODE;
 
       var message = 'Welcome to Scout! You can say, get my titles to begin.';
```
[the remark this answers is not in the document]
Comment (tamarahills) [reply state_handlers.js]: fixed.

File: state_handlers.js [hunk 10]
```diff
@@ -75,51 +91,198 @@ var state_handlers = {
...
+      skim: function() {
+        console.log('TITLES_DECISION_MODE:skim');
+        synthesisHelperUrl(this);
+      },
+      fullarticle: function() {
+        console.log('TITLES_DECISION_MODE:fullarticle');
+        synthesisHelperUrl(this);
+      },
+      SearchAndPlayArticle: function() {
+        console.log('TITLES_DECISION_MODE:SearchAndPlayArticle');
```
[the remark this answers is not in the document]
Comment (tamarahills) [reply state_handlers.js]: It has to do with the fact that when they have said they want to get the titles, they are hearing the titles so we can match them up in this case.  In the other cases where tehy just come in as "Alexa, tell scout to play that article about fake news" they have not heard the titles most likely so we call synthesisHelper which just sends a request to scout-ua and scout-ua does the matching.  This is the piece we are missing I mentioned is to either have scout-ua match against the titles with the stemming or get the titles (or keep a state flag as you mentioned).

File: state_handlers.js (hunk 1 above)
Comment (tamarahills) [reply state_handlers.js]: fixed

File: state_handlers.js (hunk 2 above)
Comment (tamarahills) [reply state_handlers.js]: Thanks.  Refactored:
``
function getTitleChunk(articleJson, stateObj) {
  let retSpeech = '<break />';

  let curCount = stateObj.attributes['titleCount'];
  let arrChunk = Object.keys(articleJson).slice(
    curCount,
    curCount + constants.TITLE_CHUNK_LEN
  );

  arrChunk.forEach(function(element, index) {
    console.log('article title: ' + arrChunk[index].title);

    retSpeech = `${retSpeech} ${index + 1}. ${arrChunk[index].title}. ${
      arrChunk[index].lengthMinutes
    } minutes.  `;
  });
  stateObj.attributes['titleCount'] += arrChunk.length;

  return retSpeech;
}

``

Comment (dylano) [reply state_handlers.js]: Cool, nice improvement. And in the foreach, can you use `element` instead of `arrChunk[index].`?

Comment (tamarahills) [reply state_handlers.js]: Yes, I realized that after I posted this, so that is in the PR that I pushed.

**Pull request merged**